{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"clustering\")\n",
    "\n",
    "import os\n",
    "from utils import get_data_dir, prinT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data_dir:  /home/lyuzhuoqi/projects/clustering/data\n",
      "data_dir:  /home/lyuzhuoqi/projects/clustering/data\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "MAG_paper_df = pd.read_parquet(os.path.join(get_data_dir(), \"MAG_paper.parquet\"))\n",
    "MAG_paper_df[MAG_paper_df['Year'] == 2010].to_parquet(os.path.join(get_data_dir(), '2010s', \"MAG_paper.parquet\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data_dir:  /home/lyuzhuoqi/projects/clustering/data\n",
      "data_dir:  /home/lyuzhuoqi/projects/clustering/data\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "MAG_paper_df = pd.read_parquet(os.path.join(get_data_dir(), \"2010s\",\"MAG_paper.parquet\"))\n",
    "cluster_df = pd.read_parquet(os.path.join(get_data_dir(), \"2010s/clustering_results\",\"cluster_df.parquet\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Year</th>\n",
       "      <th>DocType</th>\n",
       "      <th>VenueID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PaperID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2049933365</th>\n",
       "      <td>2010</td>\n",
       "      <td>Journal</td>\n",
       "      <td>80951755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2038148770</th>\n",
       "      <td>2010</td>\n",
       "      <td>Journal</td>\n",
       "      <td>119525064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2373199189</th>\n",
       "      <td>2010</td>\n",
       "      <td>Journal</td>\n",
       "      <td>2764482698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2289331308</th>\n",
       "      <td>2010</td>\n",
       "      <td>Journal</td>\n",
       "      <td>2764425571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2784227654</th>\n",
       "      <td>2010</td>\n",
       "      <td>Journal</td>\n",
       "      <td>77020770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2040927587</th>\n",
       "      <td>2010</td>\n",
       "      <td>Journal</td>\n",
       "      <td>135866459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3030134270</th>\n",
       "      <td>2010</td>\n",
       "      <td>Journal</td>\n",
       "      <td>2764740593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2087753748</th>\n",
       "      <td>2010</td>\n",
       "      <td>Journal</td>\n",
       "      <td>1567313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2144609025</th>\n",
       "      <td>2010</td>\n",
       "      <td>Journal</td>\n",
       "      <td>132415769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2405977674</th>\n",
       "      <td>2010</td>\n",
       "      <td>Journal</td>\n",
       "      <td>2752469658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3213996 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Year  DocType     VenueID\n",
       "PaperID                              \n",
       "2049933365  2010  Journal    80951755\n",
       "2038148770  2010  Journal   119525064\n",
       "2373199189  2010  Journal  2764482698\n",
       "2289331308  2010  Journal  2764425571\n",
       "2784227654  2010  Journal    77020770\n",
       "...          ...      ...         ...\n",
       "2040927587  2010  Journal   135866459\n",
       "3030134270  2010  Journal  2764740593\n",
       "2087753748  2010  Journal     1567313\n",
       "2144609025  2010  Journal   132415769\n",
       "2405977674  2010  Journal  2752469658\n",
       "\n",
       "[3213996 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MAG_paper_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Scopus_label</th>\n",
       "      <th>movMF_label</th>\n",
       "      <th>movMF_distance</th>\n",
       "      <th>x_val</th>\n",
       "      <th>y_val</th>\n",
       "      <th>kmeans_label</th>\n",
       "      <th>kmeans_distance</th>\n",
       "      <th>skm_label</th>\n",
       "      <th>skm_distance</th>\n",
       "      <th>spectral_label</th>\n",
       "      <th>n2v_kmeans_label</th>\n",
       "      <th>cm_kmeans_label</th>\n",
       "      <th>bert_kmeans_label</th>\n",
       "      <th>gnn_kmeans_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>202381698</th>\n",
       "      <td>Multidisciplinary</td>\n",
       "      <td>22</td>\n",
       "      <td>0.445886</td>\n",
       "      <td>-67.928200</td>\n",
       "      <td>15.572327</td>\n",
       "      <td>17</td>\n",
       "      <td>0.628846</td>\n",
       "      <td>20</td>\n",
       "      <td>0.444711</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137773608</th>\n",
       "      <td>Multidisciplinary</td>\n",
       "      <td>22</td>\n",
       "      <td>0.590942</td>\n",
       "      <td>-68.405334</td>\n",
       "      <td>-55.633186</td>\n",
       "      <td>17</td>\n",
       "      <td>0.735654</td>\n",
       "      <td>20</td>\n",
       "      <td>0.559494</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125754415</th>\n",
       "      <td>Multidisciplinary</td>\n",
       "      <td>22</td>\n",
       "      <td>0.574571</td>\n",
       "      <td>-68.448853</td>\n",
       "      <td>-55.613579</td>\n",
       "      <td>17</td>\n",
       "      <td>0.705024</td>\n",
       "      <td>20</td>\n",
       "      <td>0.550081</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3880285</th>\n",
       "      <td>Multidisciplinary</td>\n",
       "      <td>23</td>\n",
       "      <td>0.618842</td>\n",
       "      <td>-68.407288</td>\n",
       "      <td>-55.634430</td>\n",
       "      <td>8</td>\n",
       "      <td>0.724859</td>\n",
       "      <td>17</td>\n",
       "      <td>0.610582</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>18</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111155417</th>\n",
       "      <td>Chemistry</td>\n",
       "      <td>23</td>\n",
       "      <td>0.220853</td>\n",
       "      <td>-54.506985</td>\n",
       "      <td>-61.217068</td>\n",
       "      <td>11</td>\n",
       "      <td>0.495787</td>\n",
       "      <td>12</td>\n",
       "      <td>0.198758</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>24</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2764485818</th>\n",
       "      <td>Medicine</td>\n",
       "      <td>21</td>\n",
       "      <td>0.268070</td>\n",
       "      <td>-28.969574</td>\n",
       "      <td>34.819569</td>\n",
       "      <td>18</td>\n",
       "      <td>0.542531</td>\n",
       "      <td>14</td>\n",
       "      <td>0.251015</td>\n",
       "      <td>8</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83454320</th>\n",
       "      <td>Arts and Humanities</td>\n",
       "      <td>25</td>\n",
       "      <td>0.034777</td>\n",
       "      <td>78.609909</td>\n",
       "      <td>31.736822</td>\n",
       "      <td>13</td>\n",
       "      <td>0.251599</td>\n",
       "      <td>3</td>\n",
       "      <td>0.056642</td>\n",
       "      <td>21</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16507453</th>\n",
       "      <td>Arts and Humanities</td>\n",
       "      <td>6</td>\n",
       "      <td>0.113656</td>\n",
       "      <td>89.206772</td>\n",
       "      <td>17.625090</td>\n",
       "      <td>13</td>\n",
       "      <td>0.307089</td>\n",
       "      <td>3</td>\n",
       "      <td>0.128660</td>\n",
       "      <td>21</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121509672</th>\n",
       "      <td>Agricultural and Biological Sciences</td>\n",
       "      <td>5</td>\n",
       "      <td>0.179111</td>\n",
       "      <td>-36.757057</td>\n",
       "      <td>-0.591017</td>\n",
       "      <td>21</td>\n",
       "      <td>0.430040</td>\n",
       "      <td>16</td>\n",
       "      <td>0.207435</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53107364</th>\n",
       "      <td>Medicine</td>\n",
       "      <td>23</td>\n",
       "      <td>0.258839</td>\n",
       "      <td>-46.797390</td>\n",
       "      <td>-55.657715</td>\n",
       "      <td>11</td>\n",
       "      <td>0.494533</td>\n",
       "      <td>12</td>\n",
       "      <td>0.223036</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>23</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20038 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Scopus_label  movMF_label  movMF_distance  \\\n",
       "202381698                      Multidisciplinary           22        0.445886   \n",
       "137773608                      Multidisciplinary           22        0.590942   \n",
       "125754415                      Multidisciplinary           22        0.574571   \n",
       "3880285                        Multidisciplinary           23        0.618842   \n",
       "111155417                              Chemistry           23        0.220853   \n",
       "...                                          ...          ...             ...   \n",
       "2764485818                              Medicine           21        0.268070   \n",
       "83454320                     Arts and Humanities           25        0.034777   \n",
       "16507453                     Arts and Humanities            6        0.113656   \n",
       "121509672   Agricultural and Biological Sciences            5        0.179111   \n",
       "53107364                                Medicine           23        0.258839   \n",
       "\n",
       "                x_val      y_val  kmeans_label  kmeans_distance  skm_label  \\\n",
       "202381698  -67.928200  15.572327            17         0.628846         20   \n",
       "137773608  -68.405334 -55.633186            17         0.735654         20   \n",
       "125754415  -68.448853 -55.613579            17         0.705024         20   \n",
       "3880285    -68.407288 -55.634430             8         0.724859         17   \n",
       "111155417  -54.506985 -61.217068            11         0.495787         12   \n",
       "...               ...        ...           ...              ...        ...   \n",
       "2764485818 -28.969574  34.819569            18         0.542531         14   \n",
       "83454320    78.609909  31.736822            13         0.251599          3   \n",
       "16507453    89.206772  17.625090            13         0.307089          3   \n",
       "121509672  -36.757057  -0.591017            21         0.430040         16   \n",
       "53107364   -46.797390 -55.657715            11         0.494533         12   \n",
       "\n",
       "            skm_distance  spectral_label  n2v_kmeans_label  cm_kmeans_label  \\\n",
       "202381698       0.444711              24                 3               21   \n",
       "137773608       0.559494              24                20               21   \n",
       "125754415       0.550081              24                20               21   \n",
       "3880285         0.610582              24                20               21   \n",
       "111155417       0.198758              14                 4               24   \n",
       "...                  ...             ...               ...              ...   \n",
       "2764485818      0.251015               8                18                1   \n",
       "83454320        0.056642              21                 5                1   \n",
       "16507453        0.128660              21                 5                1   \n",
       "121509672       0.207435              10                 5                1   \n",
       "53107364        0.223036              14                 0               24   \n",
       "\n",
       "            bert_kmeans_label  gnn_kmeans_label  \n",
       "202381698                  18                 0  \n",
       "137773608                   8                10  \n",
       "125754415                   8                 0  \n",
       "3880285                    18                10  \n",
       "111155417                   6                 3  \n",
       "...                       ...               ...  \n",
       "2764485818                 23                11  \n",
       "83454320                   23                14  \n",
       "16507453                   23                14  \n",
       "121509672                  23                 1  \n",
       "53107364                   23                 4  \n",
       "\n",
       "[20038 rows x 14 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate `valid_paper.pkl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-02-11 17:09:34 -- Mapping paper ID to venue ID\n",
      "2025-02-11 17:09:35 -- Mapping venue ID to Scopus label\n",
      "2025-02-11 17:09:35 -- Mapping venue ID to movMF label\n",
      "2025-02-11 17:09:35 -- Mapping venue ID to KMeans label\n",
      "2025-02-11 17:09:35 -- Mapping venue ID to SKM label\n",
      "2025-02-11 17:09:35 -- Mapping venue ID to spectral label\n",
      "2025-02-11 17:09:35 -- Mapping venue ID to n2v_kmeans label\n",
      "2025-02-11 17:09:35 -- Mapping venue ID to cm_kmeans label\n",
      "2025-02-11 17:09:35 -- Mapping venue ID to bert_kmeans label\n",
      "2025-02-11 17:09:35 -- Mapping venue ID to gnn_kmeans label\n",
      "2025-02-11 17:09:35 -- Filtering out invalid papers\n"
     ]
    }
   ],
   "source": [
    "# 获取所有 PaperID 和对应的期刊 ID\n",
    "prinT(\"Mapping paper ID to venue ID\")\n",
    "paper_to_venue = MAG_paper_df[\"VenueID\"].to_dict()\n",
    "# 获取期刊 ID 和对应的 Scopus_label\n",
    "prinT(\"Mapping venue ID to Scopus label\")\n",
    "venue_to_Scopus_label = cluster_df[\"Scopus_label\"].to_dict()\n",
    "# 获取期刊 ID 和对应的 movMF_label\n",
    "prinT(\"Mapping venue ID to movMF label\")\n",
    "venue_to_movMF_label = cluster_df[\"movMF_label\"].to_dict()\n",
    "# 获取期刊 ID 和对应的 kmeans_label\n",
    "prinT(\"Mapping venue ID to KMeans label\")\n",
    "venue_to_kmeans_label = cluster_df[\"kmeans_label\"].to_dict()\n",
    "# 获取期刊 ID 和对应的 skm_label\n",
    "prinT(\"Mapping venue ID to SKM label\")\n",
    "venue_to_skm_label = cluster_df[\"skm_label\"].to_dict()\n",
    "# 获取期刊 ID 和对应的 spectral_label\n",
    "prinT(\"Mapping venue ID to spectral label\")\n",
    "venue_to_spectral_label = cluster_df[\"spectral_label\"].to_dict()\n",
    "# 获取期刊 ID 和对应的 n2v_kmeans_label\n",
    "prinT(\"Mapping venue ID to n2v_kmeans label\")\n",
    "venue_to_n2v_kmeans_label = cluster_df[\"n2v_kmeans_label\"].to_dict()\n",
    "# 获取期刊 ID 和对应的 cm_kmeans_label\n",
    "prinT(\"Mapping venue ID to cm_kmeans label\")\n",
    "venue_to_cm_kmeans_label = cluster_df[\"cm_kmeans_label\"].to_dict()\n",
    "# 获取期刊 ID 和对应的 bert_kmeans_label\n",
    "prinT(\"Mapping venue ID to bert_kmeans label\")\n",
    "venue_to_bert_kmeans_label = cluster_df[\"bert_kmeans_label\"].to_dict()\n",
    "# 获取期刊 ID 和对应的 gnn_kmeans_label\n",
    "prinT(\"Mapping venue ID to gnn_kmeans label\")\n",
    "venue_to_gnn_kmeans_label = cluster_df[\"gnn_kmeans_label\"].to_dict()\n",
    "\n",
    "# 过滤出有效的 PaperID 和对应 Scopus_label\n",
    "prinT(\"Filtering out invalid papers\")\n",
    "valid_papers = {'Scopus_label': {pid: venue_to_Scopus_label[vid] for pid, vid in paper_to_venue.items() if vid in \n",
    "                                 venue_to_Scopus_label},\n",
    "                'movMF_label': {pid: venue_to_movMF_label[vid] for pid, vid in paper_to_venue.items() if vid in \n",
    "                                  venue_to_movMF_label},\n",
    "                'kmeans_label': {pid: venue_to_kmeans_label[vid] for pid, vid in paper_to_venue.items() if vid in \n",
    "                                  venue_to_kmeans_label},\n",
    "                'skm_label': {pid: venue_to_skm_label[vid] for pid, vid in paper_to_venue.items() if vid in \n",
    "                                  venue_to_skm_label},\n",
    "                'spectral_label': {pid: venue_to_spectral_label[vid] for pid, vid in paper_to_venue.items() if vid in \n",
    "                                  venue_to_spectral_label},\n",
    "                'n2v_kmeans_label': {pid: venue_to_n2v_kmeans_label[vid] for pid, vid in paper_to_venue.items() if vid in \n",
    "                                  venue_to_n2v_kmeans_label},\n",
    "                'cm_kmeans_label': {pid: venue_to_cm_kmeans_label[vid] for pid, vid in paper_to_venue.items() if vid in \n",
    "                                  venue_to_cm_kmeans_label},\n",
    "                'bert_kmeans_label': {pid: venue_to_bert_kmeans_label[vid] for pid, vid in paper_to_venue.items() if vid in\n",
    "                                      venue_to_bert_kmeans_label},\n",
    "                'gnn_kmeans_label': {pid: venue_to_gnn_kmeans_label[vid] for pid, vid in paper_to_venue.items() if vid in\n",
    "                                        venue_to_gnn_kmeans_label},\n",
    "                                  }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1915871"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(valid_papers['Scopus_label'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data_dir:  /home/lyuzhuoqi/projects/clustering/data\n"
     ]
    }
   ],
   "source": [
    "with open(os.path.join(get_data_dir(), \"2010s/classification_tasks\",\"valid_papers.pkl\"), \"wb\") as f:\n",
    "    pickle.dump(valid_papers, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load `valid_paper.pkl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data_dir:  /home/lyuzhuoqi/projects/clustering/data\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "with open(os.path.join(get_data_dir(), \"2010s/classification_tasks\", \"valid_papers.pkl\"), \"rb\") as f:\n",
    "    valid_papers = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct `paper_abstract_labels.json`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ijson\n",
    "import json\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "\n",
    "def extract_valid_papers(json_path, valid_papers, output_path):\n",
    "    \"\"\"\n",
    "    从平坦的 JSON 文件（{ \"PaperID\": \"abstract\", ... }）中提取指定的 PaperID 和摘要。\n",
    "    \n",
    "    Args:\n",
    "        json_path (str): JSON 文件路径。\n",
    "        valid_papers (dict): PaperID 和其对应标签的字典。\n",
    "        output_path (str): 提取后保存的文件路径。\n",
    "    \n",
    "    Returns:\n",
    "        int: 匹配到的 PaperID 数量。\n",
    "    \"\"\"\n",
    "    matched_count = 0\n",
    "\n",
    "    # 获取文件大小用于进度条\n",
    "    file_size = os.path.getsize(json_path)\n",
    "\n",
    "    with open(json_path, 'r', encoding='utf-8') as input_file, open(output_path, 'w', encoding='utf-8') as output_file:\n",
    "        # 初始化 tqdm 进度条\n",
    "        with tqdm(total=file_size, desc=\"Processing JSON\", unit=\"B\", unit_scale=True) as pbar:\n",
    "            parser = ijson.kvitems(input_file, \"\")  # 解析顶层键值对\n",
    "            batch = []\n",
    "            batch_size = 50000  # 增大批量写入规模\n",
    "            counter = 0  # 控制进度条更新频率\n",
    "\n",
    "            for paper_id, abstract in parser:\n",
    "                counter += 1\n",
    "                if counter % 10000 == 0:  # 每 1000 条更新一次进度条\n",
    "                    pbar.update(input_file.tell() - pbar.n)\n",
    "\n",
    "                try:\n",
    "                    paper_id = int(paper_id)\n",
    "                    if paper_id in valid_papers['Scopus_label']:\n",
    "                        result = {\n",
    "                            \"PaperID\": paper_id,\n",
    "                            \"abstract\": abstract,\n",
    "                            \"Scopus_label\": valid_papers['Scopus_label'][paper_id],\n",
    "                            \"movMF_label\": valid_papers['movMF_label'][paper_id],\n",
    "                            \"kmeans_label\": valid_papers['kmeans_label'][paper_id],\n",
    "                            \"skm_label\": valid_papers['skm_label'][paper_id],\n",
    "                            \"spectral_label\": valid_papers['spectral_label'][paper_id],\n",
    "                            \"n2v_kmeans_label\": valid_papers['n2v_kmeans_label'][paper_id],\n",
    "                            \"cm_kmeans_label\": valid_papers['cm_kmeans_label'][paper_id],\n",
    "                            \"bert_kmeans_label\": valid_papers['bert_kmeans_label'][paper_id],\n",
    "                            \"gnn_kmeans_label\": valid_papers['gnn_kmeans_label'][paper_id],\n",
    "                        }\n",
    "                        batch.append(result)\n",
    "                        matched_count += 1\n",
    "                except ValueError:\n",
    "                    continue\n",
    "\n",
    "                if len(batch) >= batch_size:\n",
    "                    output_file.write(\"\\n\".join(json.dumps(item) for item in batch) + \"\\n\")\n",
    "                    batch = []\n",
    "\n",
    "            if batch:  # 写入剩余的批量数据\n",
    "                output_file.write(\"\\n\".join(json.dumps(item) for item in batch) + \"\\n\")\n",
    "\n",
    "    prinT(f\"Successfully extracted {matched_count} paper abstracts\")\n",
    "    return matched_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data_dir:  /home/lyuzhuoqi/projects/clustering/data\n",
      "data_dir:  /home/lyuzhuoqi/projects/clustering/data\n",
      "data_dir:  /home/lyuzhuoqi/projects/clustering/data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing JSON: 100%|███████████████████████████████████████████████████████▉| 177G/177G [23:12<00:00, 127MB/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-02-11 17:33:27 -- Successfully extracted 1527417 paper abstracts\n"
     ]
    }
   ],
   "source": [
    "large_json_path = os.path.join(get_data_dir(), \"MAG_abstracts.json\")\n",
    "test_json_path = os.path.join(get_data_dir(), \"MAG_abstracts_test.json\")\n",
    "output_path = os.path.join(get_data_dir(), \"2010s/classification_tasks\", \"paper_abstract_labels.json\")  # 提取后的保存路径\n",
    "\n",
    "matched_count = extract_valid_papers(large_json_path, valid_papers, output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "def read_lines(file_path, n: int=2):\n",
    "    \"\"\"\n",
    "    从 Flat JSON 文件中读取前两行并打印解析结果。\n",
    "    \n",
    "    Args:\n",
    "        file_path (str): Flat JSON 文件路径（每行一个 JSON 对象）。\n",
    "    \"\"\"\n",
    "    with open(file_path, 'r', encoding='utf-8') as f:\n",
    "        for idx, line in enumerate(f):\n",
    "            if idx < n:  # 只处理前n行\n",
    "                try:\n",
    "                    parsed_line = json.loads(line.strip())  # 解析 JSON 对象\n",
    "                    print(json.dumps(parsed_line, indent=4, ensure_ascii=False))  # 美化输出\n",
    "                except json.JSONDecodeError as e:\n",
    "                    print(f\"解析失败：{e}, 原始数据：{line.strip()}\")\n",
    "            else:\n",
    "                break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data_dir:  /home/lyuzhuoqi/projects/clustering/data\n",
      "{\n",
      "    \"PaperID\": 2055207790,\n",
      "    \"abstract\": \"Matrix metalloproteinase-13 (MMP-13), referred to as collagenase-3, is a proteolytic enzyme that plays a key role in degradation and remodelling of host extracellular matrix proteins. The objective of this study was to characterize the MMP-13 gene in channel catfish, and to determine its pattern of expression in various healthy tissues and during embryogenesis. Since MMP-13 has been shown to have importance in tissue remodelling and some pathological processes, we further studied its involvement in the defense responses of catfish after bacterial infection. The channel catfish MMP-13 cDNA contains an open reading frame of 1416 bp encoding 471 amino acids. Using RT-PCR analysis, MMP-13 was widely expressed in various health tissues. Using quantitative real-time PCR analysis, expression of MMP-13 gene was up-regulated by bacterial infection. During normal embryological development, MMP-13 expression was slightly increased in the first day post-fertilization and sharply up-regulated from 1-day post-fertilization through hatching.\",\n",
      "    \"Scopus_label\": \"Immunology and Microbiology\",\n",
      "    \"movMF_label\": 22,\n",
      "    \"kmeans_label\": 16,\n",
      "    \"skm_label\": 23,\n",
      "    \"spectral_label\": 23,\n",
      "    \"n2v_kmeans_label\": 20,\n",
      "    \"cm_kmeans_label\": 4,\n",
      "    \"bert_kmeans_label\": 14,\n",
      "    \"gnn_kmeans_label\": 23\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055207847,\n",
      "    \"abstract\": \"Abstract     A technique to generate random fractal aggregates where the fractal dimension is fixed a priori is presented. The algorithm utilizes the box-counting measure of the fractal dimension to determine the number of hypercubes required to encompass the aggregate, on a set of length scales, over which the structure can be defined as fractal. At each length scale the hypercubes required to generate the structure are chosen using a simple random walk which ensures connectivity of the aggregate. The algorithm is highly efficient and overcomes the limitations on the magnitude of the fractal dimension encountered by previous techniques.\",\n",
      "    \"Scopus_label\": \"Physics and Astronomy\",\n",
      "    \"movMF_label\": 16,\n",
      "    \"kmeans_label\": 15,\n",
      "    \"skm_label\": 0,\n",
      "    \"spectral_label\": 20,\n",
      "    \"n2v_kmeans_label\": 16,\n",
      "    \"cm_kmeans_label\": 14,\n",
      "    \"bert_kmeans_label\": 18,\n",
      "    \"gnn_kmeans_label\": 10\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055207981,\n",
      "    \"abstract\": \"Protein phosphorylation is one of the most common and extensively studied posttranslational modifications (PTMs). Compared to the O-phosphorylation of Ser, Thr, and Tyr residues, our understanding of histidine phosphorylation is relatively limited, particularly in higher eukaryotes, due to technical difficulties stemming from the intrinsic instability and isomerism of phosphohistidine (pHis). We report the design and synthesis of stable and nonisomerizable pHis analogues. These pHis analogues were successfully utilized in solid-phase peptide synthesis and semi-synthesis of histone H4. Significantly, the first antibody that specifically recognizes pHis was obtained using the synthetic peptide as the immunogen.\",\n",
      "    \"Scopus_label\": \"Chemistry\",\n",
      "    \"movMF_label\": 23,\n",
      "    \"kmeans_label\": 11,\n",
      "    \"skm_label\": 12,\n",
      "    \"spectral_label\": 14,\n",
      "    \"n2v_kmeans_label\": 4,\n",
      "    \"cm_kmeans_label\": 24,\n",
      "    \"bert_kmeans_label\": 6,\n",
      "    \"gnn_kmeans_label\": 3\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055208080,\n",
      "    \"abstract\": \"Andrew Heaney has been executive director of strategy and regulation with TalkTalk since 2007. He was previously a competition policy director with Ofcom and a partner with Spectrum Strategy Consultants.\",\n",
      "    \"Scopus_label\": \"Social Sciences\",\n",
      "    \"movMF_label\": 2,\n",
      "    \"kmeans_label\": 24,\n",
      "    \"skm_label\": 4,\n",
      "    \"spectral_label\": 2,\n",
      "    \"n2v_kmeans_label\": 19,\n",
      "    \"cm_kmeans_label\": 16,\n",
      "    \"bert_kmeans_label\": 5,\n",
      "    \"gnn_kmeans_label\": 25\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055208198,\n",
      "    \"abstract\": \"Performance measurement has become an established means by which local governments worldwide strive to improve public services. In the United States of America (USA), performance measurement is frequently combined with benchmarking. Benchmarking presupposes comparability of services and performance metrics. When each local government is free to create its own performance measurement system and select its own performance metrics, the result can be a 'Tower of Babel' that precludes valid benchmarking. This article presents a case study of the Florida Benchmarking Consortium (FBC), a group of some 48 local governments that have come together in an attempt to develop common service definitions and metrics in order to facilitate benchmarking.\",\n",
      "    \"Scopus_label\": \"Social Sciences\",\n",
      "    \"movMF_label\": 2,\n",
      "    \"kmeans_label\": 24,\n",
      "    \"skm_label\": 4,\n",
      "    \"spectral_label\": 6,\n",
      "    \"n2v_kmeans_label\": 0,\n",
      "    \"cm_kmeans_label\": 1,\n",
      "    \"bert_kmeans_label\": 7,\n",
      "    \"gnn_kmeans_label\": 2\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055208235,\n",
      "    \"abstract\": \"To determine the feasibility and efficacy of stereotactic body radiotherapy (SBRT) for oligometastases from colorectal cancer (CRC). Total of 59 patients with 78 lesions confined to one organ and treated from 2001 to 2006 were involved in this retrospective review. These patients presented with 1-4 metastatic lesions of largest diameter of <7 cm, progressed after chemotherapy, and were amenable to local treatment. The median radiation dose administered was 42 Gy, which was delivered in 3 fractions. Lymph node lesions were most frequent, followed in order by the lung and liver. Five-year overall survival and local control rates were 29 and 19%, respectively. Cumulative gross tumor volume, site of metastasis, and SBRT dose were found to be significantly associated with overall survival. In terms of local control, a cumulative GTV below 23 ml was found to be a significantly favorable prognostic factor. Acute grade 1-2 toxicities occurred in 24 of the 59 patients, and a late grade 4 complication occurred in 2 (3%), and these were relieved by bypass surgery. CRC patients with oligometastases generally fare well after SBRT. In particular, the survival rates of patients with a pelvic LN or a small tumor are promising.\",\n",
      "    \"Scopus_label\": \"Medicine\",\n",
      "    \"movMF_label\": 20,\n",
      "    \"kmeans_label\": 4,\n",
      "    \"skm_label\": 20,\n",
      "    \"spectral_label\": 13,\n",
      "    \"n2v_kmeans_label\": 20,\n",
      "    \"cm_kmeans_label\": 3,\n",
      "    \"bert_kmeans_label\": 8,\n",
      "    \"gnn_kmeans_label\": 0\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055208283,\n",
      "    \"abstract\": \"Leptin can enhance thymopoiesis and modulate the T-cell immune response. However, it remains controversial whether these effects correlate with the expression of leptin receptor, ObR. We herein addressed this issue by using in vivo animal models and in vitro culture systems. Leptin treatment in both ob/ob mice and normal young mice induced increases of CD4 SP thymocytes in thymus and CD4 T cells in the periphery. Interestingly, expression of the long form ObR was significantly restricted to DN, DP and CD4 SP, but not CD8 SP thymocytes. Moreover, in the reaggregated DP thymocyte cultures with leptin plus TSCs, leptin profoundly induced differentiation of CD4 SP but not CD8 SP thymocytes, suggesting that the effects of leptin on thymocyte differentiation might be closely related to the expression of leptin receptor in developing thymocytes. Surprisingly, ObR expression was markedly higher in peripheral CD4 T cells than that in CD8 T cells. Furthermore, leptin treatment with or without IL-2 and PHA had preferential effects on cell proliferation of CD4 T cells compared to that of CD8 T cells. Collectively, these data provide evidence that the effects of leptin on differentiation and proliferation of CD4 T cells might be closely related to the expression of leptin receptor.\",\n",
      "    \"Scopus_label\": \"Biochemistry, Genetics and Molecular Biology\",\n",
      "    \"movMF_label\": 22,\n",
      "    \"kmeans_label\": 17,\n",
      "    \"skm_label\": 20,\n",
      "    \"spectral_label\": 24,\n",
      "    \"n2v_kmeans_label\": 20,\n",
      "    \"cm_kmeans_label\": 21,\n",
      "    \"bert_kmeans_label\": 8,\n",
      "    \"gnn_kmeans_label\": 10\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055208499,\n",
      "    \"abstract\": \"Separation into spectral and nonlinear complex-eigenvalue problems is shown to be an effective and flexible approach to soliton laser models. The simplest such model, a complex Ginzburg-Landau model with cubic nonlinearity, has no stable solitonic solutions. We show that coupling it to a resonant linear system is a simple and general route to stabilization, which encompasses several previous instances in both space- and time-domains. Graphical solution in the complex eigenvalue plane provides valuable insight into the similarities and differences of such models, and into the interpretation of related experiments. It can also be used predictively, to guide analysis, numerics and experiment.\",\n",
      "    \"Scopus_label\": \"Physics and Astronomy\",\n",
      "    \"movMF_label\": 23,\n",
      "    \"kmeans_label\": 6,\n",
      "    \"skm_label\": 10,\n",
      "    \"spectral_label\": 19,\n",
      "    \"n2v_kmeans_label\": 7,\n",
      "    \"cm_kmeans_label\": 14,\n",
      "    \"bert_kmeans_label\": 25,\n",
      "    \"gnn_kmeans_label\": 3\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055208732,\n",
      "    \"abstract\": \"Abstract     Small bioactive peptides, with diverse biological functions, have received increasing attention as physiologically beneficial substances in animal production. The main obstacle to wide application of small bioactive peptides is a lack of cost-effective methods for mass production. In this study, we used glycyl-glutamine (Gly-Gln) as a test case to develop a novel mass-production method for small bioactive peptides. The oligonucleotide encoding Gly-Gln pro-peptide (Gpp) was designed and synthesized. Gpp includes 3 Gly-Gln dipeptides and 2 enzymatic sites for pepsin and trypase, allowing direct digestion and absorption of Gly-Gln in the gastrointestinal tract. The Gpp oligonucleotides were linked to generate an oligomeric oligonucleotide segment containing 12 tandem copies of Gpp. This 12Gpp segment was cloned and expressed in   Escherichia coli   vector pET32a. By optimizing culture conditions [0.1 mmol L −1   isopropyl-β-D-thiogalactopyranoside (IPTG), 50 μg mL −1   ampicillin (Amp), 30°C for 12 h], the thioredoxin fusion peptides reached 40% of total bacterial protein. After purification, the fusion protein was fed to Kunming mice to determine its effect on mouse immune function. The results showed that similar to Gly-Gln dipeptide, Gpp polymer protein could significantly suppress the proliferation of T and B lymphocytes in blood and spleen, and additionally could significantly improve interleukin-2 (IL-2) and interleukin-6 (IL-6) secretion of blood and spleen lymphocytes. These effects were not observed in mice fed a 2 amino acids mix (glycine and glutamine). These evidences indicated that an efficient digestion of Gpp polymer protein could be achieved when ingested into the animal gut. The expression system in this study provides a potential production method for not only Gly-Gln dipeptide but also other short bioactive peptides.\",\n",
      "    \"Scopus_label\": \"Agricultural and Biological Sciences\",\n",
      "    \"movMF_label\": 5,\n",
      "    \"kmeans_label\": 21,\n",
      "    \"skm_label\": 16,\n",
      "    \"spectral_label\": 10,\n",
      "    \"n2v_kmeans_label\": 14,\n",
      "    \"cm_kmeans_label\": 4,\n",
      "    \"bert_kmeans_label\": 14,\n",
      "    \"gnn_kmeans_label\": 5\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055208918,\n",
      "    \"abstract\": \"This article illustrates the challenges practitioners face evaluating shortness of breath, a common emergency department complaint. Through a series of patient encounters, pitfalls in the evaluation of shortness of breath are reviewed and discussed.\",\n",
      "    \"Scopus_label\": \"Medicine\",\n",
      "    \"movMF_label\": 15,\n",
      "    \"kmeans_label\": 7,\n",
      "    \"skm_label\": 19,\n",
      "    \"spectral_label\": 5,\n",
      "    \"n2v_kmeans_label\": 3,\n",
      "    \"cm_kmeans_label\": 5,\n",
      "    \"bert_kmeans_label\": 20,\n",
      "    \"gnn_kmeans_label\": 11\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055209225,\n",
      "    \"abstract\": \"In this paper we will address the problem of the existence of orbitals by analyzing the relationship between molecular chemistry and quantum mechanics. In particular, we will consider the concept of orbital in the light of the arguments that deny its referring character. On this basis, we will conclude that the claim that orbitals do not exist relies on a metaphysical reductionism which, if consistently sustained, would lead to consequences clashing with the effective practice of science in its different branches.\",\n",
      "    \"Scopus_label\": \"Arts and Humanities\",\n",
      "    \"movMF_label\": 24,\n",
      "    \"kmeans_label\": 11,\n",
      "    \"skm_label\": 3,\n",
      "    \"spectral_label\": 14,\n",
      "    \"n2v_kmeans_label\": 14,\n",
      "    \"cm_kmeans_label\": 1,\n",
      "    \"bert_kmeans_label\": 0,\n",
      "    \"gnn_kmeans_label\": 2\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055209304,\n",
      "    \"abstract\": \"Abstract Background: Hospice and palliative care organizations are expanding their use of standardized instruments and other approaches to measure quality. We undertook a systematic review and evaluation of published patient-level instruments for potential application in hospice and palliative care clinical quality measurement. Methods: We searched prior reviews and computerized reference databases from 1990 through February 2007 for studies of instruments relevant to physical, psychological, social, cultural, spiritual, or ethical aspects of palliative care, or measuring prognosis, function or continuity of care. Publications were selected for full review if they provided evidence of psychometric properties or practical application of an instrument tested in or appropriate for a hospice or palliative care population. Selected instruments were evaluated and scored for scientific soundness and potential application in clinical quality measurement. Results: The search found 1427 publications, with 229 selec...\",\n",
      "    \"Scopus_label\": \"Medicine\",\n",
      "    \"movMF_label\": 12,\n",
      "    \"kmeans_label\": 20,\n",
      "    \"skm_label\": 1,\n",
      "    \"spectral_label\": 5,\n",
      "    \"n2v_kmeans_label\": 1,\n",
      "    \"cm_kmeans_label\": 18,\n",
      "    \"bert_kmeans_label\": 20,\n",
      "    \"gnn_kmeans_label\": 11\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055209420,\n",
      "    \"abstract\": \"Metastasis of cancer cells is the main cause of death in most breast cancer patients. Although markers for early diagnosis and drugs that limit the spread of cancer to other organs have been developed, it is difficult to prevent the relapse of breast cancer. Recent research has highlighted the importance of tumor environment in which communication between tumor cells and the body system occurs. Emerging data have suggested that animal models are a good system to investigate this communication. Therefore, studies with mouse models have been developed as a reasonable method for a systemic approach to understand breast cancer metastasis. In this review, we summarize mouse models of breast cancer and their applications to the study of human breast cancers, and discuss limitation of model system and advanced techniques to overcome it.\",\n",
      "    \"Scopus_label\": \"Biochemistry, Genetics and Molecular Biology\",\n",
      "    \"movMF_label\": 22,\n",
      "    \"kmeans_label\": 17,\n",
      "    \"skm_label\": 20,\n",
      "    \"spectral_label\": 24,\n",
      "    \"n2v_kmeans_label\": 20,\n",
      "    \"cm_kmeans_label\": 21,\n",
      "    \"bert_kmeans_label\": 8,\n",
      "    \"gnn_kmeans_label\": 10\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055209464,\n",
      "    \"abstract\": \"Abstract     Plasma-sprayed yttria-stabilized zirconia particles (∼ 40 µm diameter) were photographed during impact (velocity ∼ 200 m/s) on glass surfaces that were maintained at either room temperature or 400 °C. A droplet that approached the surface was sensed using a photodetector and after a known delay, a light source was triggered to illuminate the particle in order to photograph it with a charge-coupled device (CCD) camera. A rapid two-color pyrometer was used to collect the thermal radiation from the particles to follow the evolution of their temperature and size, in-flight and after impact. The fully molten particles spread into a thin liquid splat after impacting the surfaces. The partially molten particles disintegrated into small satellite fragments immediately upon impact. The surface area, as indicated by the pyrometric signals, of the partially molten particles during spreading was almost an order of magnitude smaller than that of the fully molten droplets. The pyrometric signals of partially molten zirconia showed significant fluctuations, lower maximum voltages, and were wider than those of fully molten zirconia. These signals, characteristic of the impact of partially molten zirconia, provide a novel method of identifying partially molten ceramic particles after impact on a flat surface.\",\n",
      "    \"Scopus_label\": \"Physics and Astronomy\",\n",
      "    \"movMF_label\": 23,\n",
      "    \"kmeans_label\": 14,\n",
      "    \"skm_label\": 12,\n",
      "    \"spectral_label\": 17,\n",
      "    \"n2v_kmeans_label\": 4,\n",
      "    \"cm_kmeans_label\": 6,\n",
      "    \"bert_kmeans_label\": 6,\n",
      "    \"gnn_kmeans_label\": 3\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055209554,\n",
      "    \"abstract\": \"Objective   To assess the therapeutic effect and adverse reaction of   Qufeng Zhidong   Recipe (a recipe for dispelling wind to stop abnormal movement) used to treat children with tic disorder (TD).       Methods   The enrolled patients were randomized into a TCM group (31 cases) treated with   Qufeng Zhidong   Recipe and a Western medicine group (30 cases) treated with haloperidol and trihexyphenidyl. Two courses of treatment were observed with 12 weeks as one course. The therapeutic effect and adverse reaction were assessed with Yale Global Tic Severity Scale (YGTSS), Tic Symptom Score Scale (TSSS), TCM Syndrome Score Scale (TCMSSS), Treatment Emergent Symptom Scale (TESS) and laboratory examinations.       Results   The total effective rate was 100% in the TCM group and 60% in the Western medicine group with statistical significance in difference ( P   P         Conclusion     Qufeng Zhidong   Recipe can obviously relieve the symptoms and signs of TD children without toxic side-effects.\",\n",
      "    \"Scopus_label\": \"Medicine\",\n",
      "    \"movMF_label\": 22,\n",
      "    \"kmeans_label\": 17,\n",
      "    \"skm_label\": 6,\n",
      "    \"spectral_label\": 24,\n",
      "    \"n2v_kmeans_label\": 10,\n",
      "    \"cm_kmeans_label\": 4,\n",
      "    \"bert_kmeans_label\": 2,\n",
      "    \"gnn_kmeans_label\": 1\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055209575,\n",
      "    \"abstract\": \"We propose PWC, a proactive worm containment solution for enterprises. PWC can stop—instead of just slow down—an infected host from releasing worm scans as early as after merely four scans. Motivated by the observation that a worm uses a sustained outgoing packet rate, PWC gains infection awareness seconds before a signature or filter can be generated. To overcome denial-of-service possibly caused by such characteristic indicators of infection, PWC/,develops two new white detection (detecting who are uninfected) techniques: (a) the vulnerability time window lemma, and (b) the relaxation analysis. PWC does not rely on contents-based signatures thus it can defend against polymorphic worms timely in containment. PWC is also resilient to containment evading. PWC is not sensitive to worm scan rate, and not protocol specific. Due to white detection, PWC causes minimal denial-of-service. Evaluation based on real traces and worm simulations demonstrates that PWC significantly outperforms Virus Throttle in terms of number of released worm scans, number of hosts infected by local scans, and denial-of-service effects. Copyright © 2009 John Wiley & Sons, Ltd.\",\n",
      "    \"Scopus_label\": \"Computer Science\",\n",
      "    \"movMF_label\": 14,\n",
      "    \"kmeans_label\": 2,\n",
      "    \"skm_label\": 21,\n",
      "    \"spectral_label\": 1,\n",
      "    \"n2v_kmeans_label\": 6,\n",
      "    \"cm_kmeans_label\": 0,\n",
      "    \"bert_kmeans_label\": 18,\n",
      "    \"gnn_kmeans_label\": 25\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055209600,\n",
      "    \"abstract\": \"Nattokinase was produced by batch and fed-batch culture of   Bacillus subtilis   in flask and fermentor. Effect of supplementing complex media (peptone, yeast extract, or tryptone) was investigated on the production of nattokinase. In flask culture, the highest cell growth and nattokinase activity were obtained with 50 g/L of peptone supplementation. In this condition, nattokinase activity was 630 unit/ml at 12 h. In batch culture of   B. subtilis   in fermentor, the highest nattokinase activity of 3400 unit/ml was obtained at 10 h with 50 g/L of peptone supplementation. From the batch kinetics data, it was shown that nattokinase production was growth-associated and culture should be harvested before stationary phase for maximum nattokinase production. In fed-batch culture of   B. subtilis   using pH-stat feeding strategy, cell growth (optical density monitored at 600 nm) increased to ca. 100 at 22 h, which was 2.5 times higher than that in batch culture. The highest nattokinase activity was 7100 unit/ml at 19 h, which was also 2.1 times higher than that in batch culture.\",\n",
      "    \"Scopus_label\": \"Biochemistry, Genetics and Molecular Biology\",\n",
      "    \"movMF_label\": 22,\n",
      "    \"kmeans_label\": 21,\n",
      "    \"skm_label\": 16,\n",
      "    \"spectral_label\": 10,\n",
      "    \"n2v_kmeans_label\": 20,\n",
      "    \"cm_kmeans_label\": 21,\n",
      "    \"bert_kmeans_label\": 14,\n",
      "    \"gnn_kmeans_label\": 10\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055210024,\n",
      "    \"abstract\": \"The electric field in the neighborhood of the tip of conical point exhibits an integrable special feature. Therefore, for calculating the current of thermionic-field emission from a conical point, one has to find (or reasonably estimate) the emission current density, including that at high and, in the limit, infinite fields. It is suggested to solve this problem using, firstly, a numerical method of finding the electric field of a conical point which is capable of correctly describing the feature at the tip and, secondly, a common approach to considering the thermionic, field, and thermionic-field emission of electrons from metal in a wide range of variation of electric field on the surface. It is demonstrated that, at high electric fields, the density of thermionic-field emission current tends to some limiting value, i.e., the effect of saturation of emission current density is observed. The basic reason for this saturation is found. It is demonstrated that the total current of electron emission from a conical point remains finite at not too high values of intensity of external electric field, in spite of the infinite field increasing at the cone tip.\",\n",
      "    \"Scopus_label\": \"Physics and Astronomy\",\n",
      "    \"movMF_label\": 23,\n",
      "    \"kmeans_label\": 6,\n",
      "    \"skm_label\": 12,\n",
      "    \"spectral_label\": 17,\n",
      "    \"n2v_kmeans_label\": 7,\n",
      "    \"cm_kmeans_label\": 1,\n",
      "    \"bert_kmeans_label\": 25,\n",
      "    \"gnn_kmeans_label\": 13\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055210094,\n",
      "    \"abstract\": \"Two different types of Diels-Alder additions to morphinan-6, 8-dienes have been found. 6-Demethoxythebaine (2) yielded ethyl 4, 5α-epoxy-3-methoxy-N-methyl-6, 14-ethenoisomorphinan–7α-carboxylate (3) with ethyl acrylate, in analogy to the reaction with thebaine. The ester 3 was converted into the alcohol 4, of which the 3-methoxy ether was hydrolyzed to yield 5. Similarly, 2 gave the 7α-acetyl-6, 14-ethenoisomorphinan 7 with methyl vinyl ketone. The latter compound was converted into 4 using methylmagnesium iodide. With propylmagnesium bromide, 7 afforded four compounds; two are new etorphine analogues (8 and 9), to which we were able to assign the absolute configuration; the other two are the Grignard reduction products 10 and 11.\\r\\n\\r\\n\\r\\n\\r\\nWhen the 4, 5α-epoxy bridge of 2 is first opened, the Diels-Alder reaction with methyl vinyl ketone proceeds from the other side of the diene system yielding the 7β-acetyl-6, 14-ethenomorphinan 6, which belongs to a novel class of rigid morphinans.\\r\\n\\r\\n\\r\\n\\r\\nPreliminary pharmacological screening of 5, 8 and 9 showed these compounds to be potent agonists.\",\n",
      "    \"Scopus_label\": \"Chemistry\",\n",
      "    \"movMF_label\": 23,\n",
      "    \"kmeans_label\": 11,\n",
      "    \"skm_label\": 12,\n",
      "    \"spectral_label\": 14,\n",
      "    \"n2v_kmeans_label\": 4,\n",
      "    \"cm_kmeans_label\": 1,\n",
      "    \"bert_kmeans_label\": 6,\n",
      "    \"gnn_kmeans_label\": 3\n",
      "}\n",
      "{\n",
      "    \"PaperID\": 2055210327,\n",
      "    \"abstract\": \"We report quantum coherence measurements of a superconducting qubit whose design is a hybrid of several existing types. Excellent coherence times are found: ${T}_{2}^{*}\\\\ensuremath{\\\\sim}{T}_{1}\\\\ensuremath{\\\\sim}1.5\\\\text{ }\\\\text{ }\\\\ensuremath{\\\\mu}\\\\mathrm{s}$. The topology of the qubit is that of a traditional three-junction flux qubit, but it has a large shunting capacitance, and the ratio of the junction critical currents is chosen so that the qubit potential has a single-well form. The qubit has a sizable nonlinearity, but its sign is reversed compared with most other popular qubit designs. The qubit is read out dispersively using a high-$Q$ resonator in a $\\\\ensuremath{\\\\lambda}/2$ configuration.\",\n",
      "    \"Scopus_label\": \"Physics and Astronomy\",\n",
      "    \"movMF_label\": 23,\n",
      "    \"kmeans_label\": 6,\n",
      "    \"skm_label\": 0,\n",
      "    \"spectral_label\": 19,\n",
      "    \"n2v_kmeans_label\": 16,\n",
      "    \"cm_kmeans_label\": 14,\n",
      "    \"bert_kmeans_label\": 6,\n",
      "    \"gnn_kmeans_label\": 24\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "output_path = os.path.join(get_data_dir(), \"2010s/classification_tasks\", \"paper_abstract_labels.json\")\n",
    "read_lines(output_path, 20)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
